Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js

```diff
@@ -113,6 +115,22 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
     get locked() { return this._locked; }
     set locked(value) { this._locked = value; }
 
+    get useVariablesGrouping() { return this._useVariablesGrouping; }
+    set useVariablesGrouping(value)
+    {
+        let wasUngrouped = !this._useVariablesGrouping && value;
+        this._useVariablesGrouping = value;
+
+        // Force an update with the same parameters to trigger variable grouping.
+        if (wasUngrouped)
+            this.update(this.text, this.properties, this.styleSheetTextRange, {dontFireEvents: true});
```

Comment:
Why is this necessary?  Could we instead generate this lazily inside `variablesForType` and then clear the `Map` inside `update`?

_reads commit message_

I dont really think the performance of iterating `_properties` is likely to be of concern as I'd be very surprised if there were more than a thousand (and probably not even anything close to that) CSS variables at the same time.  Not to mention, we're effectively forcing a second iteration when switching to the Computed panel anyways, so I'm not sure what much we gain by doing this.

Reply:
> Could we instead generate this lazily inside variablesForType 

Yes. That's a good idea. Updated.

> I dont really think the performance of iterating _properties is likely to be of concern

`DOMNodeStyles.computedStyle.update()` is [called every time](https://sourcegraph.com/github.com/WebKit/WebKit-http@817c46e152af795d735678386db68805d0aa505e/-/blob/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js?L248) there's a change to the styles of the selected node. Typing a CSS property name or value means a flurry of activity. It's enough to do just one iteration over the list of properties in the computed style. 

I'd argue we could throttle or debounce the handler for the `WI.CSSStyleDeclaration.Event.PropertiesChanged` event. The Computed style panel doesn't need to be _live_ in sync with the Styles panel.

What concerns me more is that every single style change thrashes the DOM in`WI.ComputedStyleSection.layout()` by [removing all elements](https://sourcegraph.com/github.com/WebKit/WebKit-http@817c46e152af795d735678386db68805d0aa505e/-/blob/Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js?L197) and rebuilding from scratch. This is very wasteful. I'd like to introduce a `WeakMap` that matches incoming properties with previously created `WI.SpreadsheetStyleProperty` instances, use it to check what changed, and update the DOM accordingly. But that's going to happen in a follow-up. This patch is already sprawling.

Reply:
apologies, i dont think i was entirely clear

what i was really trying to get at is why are we doing this inside `update` at all?  why not just iterate over `_properties` inside `variablesForType`?
```
variablesForType(type)
{
    if (!this._variablesForType) {
        this._variablesForType = new Map;
        for (let property of this._properties) {
            if (property.isVariable)
                continue;

            ...
        } 
    }
    return this._variablesForType;
}
```
(and we'd clear `this._variablesForType` inside `update`)

calling `update` feels a bit heavy cause it also adjusts things like `this._propertyNameMap` and the `WI.CSSProperty.prototype.get index`, which IMO is entirely unnecessary

as far as the other things, i definitely agree that we should improve (in a followup) how we handle changes to the computed style :)  throttling would be a nice start for sure

Reply:
Thought about it some more and I agree. The separation wasn't necessary.
Changed the patch to do variable grouping lazily in `this.variablesForType()`.
Dropped `useVariablesGrouping`.

Now, we'll generate a `Map` with lists of variables for every group type. Empty lists for groups with no matches serve as a signal for `WI.ComputedStyleDetailsPanel` to avoid rendering sections that won't have content. This shaves some more unnecessary loops for sections known to be empty.

Reply:
Modified `WI.View` to call `WI.View.didLayoutSubtree()` on the immediate parent
of the first dirty view that did layout on itself and all of its descendant views.

Reply:
Used @dcrousso's patch to modify `WI.View._visitViewTreeForLayout()`.